### PR TITLE
Convert the WellSqlConfig migrations to Kotlin

### PR DESCRIPTION
**Pull request by 0nko**
After fixing #1393, I've been thinking about how to make the migrations safer and less error-prone. I've converted the `WellSqlConfig` class to Kotlin and extracted some repetitive code, which should help. What do you think?

I've copied over the SQL scripts but just to be safe, I'd like to ask the reviewer to double check that everything correct.

### CODE REVIEW

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt [hunk 1]
```diff
@@ -0,0 +1,838 @@
...
+                34 -> migrate(version) {
+                    db.execSQL("DROP TABLE IF EXISTS RewindStatus")
+                    db.execSQL(
+                            "CREATE TABLE RewindStatus (_id INTEGER PRIMARY KEY AUTOINCREMENT,LOCAL_SITE_ID INTEGER,"
+                                    + "REMOTE_SITE_ID INTEGER,STATE TEXT NOT NULL,LAST_UPDATED INTEGER,REASON TEXT,"
+                                    + "CAN_AUTOCONFIGURE INTEGER,REWIND_ID TEXT,REWIND_STATUS TEXT,REWIND_PROGRESS INTEGER,"
+                                    + "REWIND_REASON TEXT)"
+                    )
+                }
+                35 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
```
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt]: I think this step of the migration is missing and probably should have this query: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1395/files#diff-7b4c47eb749776091657a82024f162ceL699

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt [hunk 2]
```diff
@@ -0,0 +1,838 @@
...
+                    db.execSQL("ALTER TABLE PostModel ADD LAST_MODIFIED TEXT")
+                }
+                44 -> migrate(version) {
+                    db.execSQL("DROP TABLE IF EXISTS StatsBlock")
+                    db.execSQL(
+                            "CREATE TABLE StatsBlock (_id INTEGER PRIMARY KEY AUTOINCREMENT,LOCAL_SITE_ID INTEGER,"
+                                    + "BLOCK_TYPE TEXT NOT NULL,STATS_TYPE TEXT NOT NULL,JSON TEXT NOT NULL)"
+                    )
+                }
+                45 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
```
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt]: Would it be better to have the outer call as `migrate` and the inner call as `migrateAddOn`? Probably not very necessary, but I was just thinking in terms of the logging order. Although it's very unlikely this migration will run anymore, so there is that.

**Comment by aforcier** [comment]
Just came by to mention that the unit tests and ktlint were failing but it looks like you fixed them 👍 

Nice Kotlinification, this much improves the class 🎉.

Other than lint/tests and Oguz's notes this looks good. One thing is that the checkstyle plugin is unhappy with this file and riddles it with newline warnings in the IDE (the actual gradle task passes). Are you seeing that too?

Also for testing, I was wondering @0nko @oguzkocer if you've tried running a migration? Something like:

1. Run really old build of WPAndroid and WooAndroid
2. Login, fetch some posts, etc
3. Build current version of each app using this PR's FluxC hash
4. Run app, check the log for long series of migrations as expected
5. Make sure everything still works

It's probably fine but let's be extra cautious with our migration scripts. I'll try to try this today, if not it'll be next week.

**Comment by oguzkocer** [comment]
@aforcier I think that's a good idea. I hope the old builds will work fine with the new hash, I know it should, but with Gradle it's not always the case.

@0nko I see some new commits, can you let me know when it's ready for another pass/test?

**Comment by aforcier** [comment]
> @aforcier I think that's a good idea. I hope the old builds will work fine with the new hash, I know it should, but with Gradle it's not always the case.

Right - my suggestion was to use [old releases](https://github.com/wordpress-mobile/WordPress-Android/releases), that way they're already compiled and we don't have to worry about being able to build them. Downside is not debuggable, but that's only an issue if there turns out to be an actual problem 😀  (then we'd need to dig in)

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt (hunk 2 above)
Comment (0nko) [reply fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt]: Makes sense. I put them in sequence now.

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt (hunk 1 above)
Comment (0nko) [reply fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt]: Thanks!

**Comment by 0nko** [comment]
@aforcier @oguzkocer Thanks for the review!

I saw the checkstyle errors but I suppressed them in AS somehow and I can't seem to bring them back.

I've spent half a day yesterday trying to make the migration run from version 1 work. I found a point where the `WellSqlConfig` was just added and made it build: `0nko/fluxc-db-v-1`. There were a couple of missing create table scripts but it works fine now. It was a bit tricky, though, because for each I needed a version of the table in a specific point in history.

I also wanted to write a unit test that would run the migration the same way but I haven't figured it out yet.

**Comment by 0nko** [comment]
@aforcier, would you mind taking another look? I think it should be good, but just in case. :)

**Comment by aforcier** [comment]
I tested migrating (to a build using the tip of this branch) from a few older WPAndroid versions and one WooAndroid version:

- WPAndroid 7.0, DB version 4
- WPAndroid 9.6, DB version 25
- WooAndroid 1.1, DB version 48

In all cases the migrations ran and were logged correctly, and drafts/data seemed to be preserved.

There's a small problem though - this currently is a breaking change for WPAndroid (in debug but not in release) and WooAndroid (any build). The breaking classes are [WPWellSqlConfig](https://github.com/wordpress-mobile/WordPress-Android/blob/4c3cdaa758e85d1a6fd54b62bc7dac749916360f/WordPress/src/debug/java/org/wordpress/android/WPWellSqlConfig.kt) and [WooWellSqlConfig](https://github.com/woocommerce/woocommerce-android/blob/54273f37cd299b04ed5776ae20d7128b9c866912/WooCommerce/src/main/kotlin/com/woocommerce/android/WooWellSqlConfig.kt). There are two causes:

1. The constructors now require `Context` and a `Context?` is being supplied.
2. The second `reset()` function that accepted a `helper: WellTableManager` was removed

**Comment by oguzkocer** [comment]
Ah yes, I saw those breaks as well in WPAndroid, but didn't think it would be big deal. I should have mentioned it though since we want to have PRs ready to be merged when this change is merged to `develop`. Thanks for bringing it up @aforcier!

**Comment by 0nko** [comment]
Thanks for pointing it out, Alex! I added the original reset method with the helper parameter, which was needed for the downgrades. I've created a PR in each repo, ready for review/merge.

**Comment by aforcier** [comment]
Looks good and so do the counterpart PRs - merging this and leaving the Woo and WP PRs open so you can update the tag before those merge.

:shipit:

**Pull request merged**